plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -545,6 +555,159 @@ public ColumnHandle getDeleteRowIdColumnHandle(ConnectorSession session, Connect
+                OPTIMIZE,
+                new IcebergOptimizeHandle(
+                        SchemaParser.toJson(icebergTable.schema()),
+                        PartitionSpecParser.toJson(icebergTable.spec()),
+                        getColumns(icebergTable.schema(), typeManager),
+                        getFileFormat(icebergTable),
+                        icebergTable.properties(),
+                        maxScannedFileSize,
+                        Optional.empty()),
+                icebergTable.location()));
+    }
+
+    @Override
+    public Optional<ConnectorNewTableLayout> getLayoutForTableExecute(ConnectorSession session, ConnectorTableExecuteHandle tableExecuteHandle)
+    {
+        IcebergTableExecuteHandle executeHandle = (IcebergTableExecuteHandle) tableExecuteHandle;
+        switch (executeHandle.getProcedureId()) {
+            case OPTIMIZE:
+                return getLayoutForOptimize(session, executeHandle);
+        }
+        throw new IllegalArgumentException("Unknown procedure '" + executeHandle.getProcedureId() + "'");
+    }
+
+    private Optional<ConnectorNewTableLayout> getLayoutForOptimize(ConnectorSession session, IcebergTableExecuteHandle executeHandle)
+    {
+        Table icebergTable = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        return getWriteLayout(icebergTable.schema(), icebergTable.spec());
+    }
+
+    @Override
+    public BeginTableExecuteResult<ConnectorTableExecuteHandle, ConnectorTableHandle> beginTableExecute(
+            ConnectorSession session,
+            ConnectorTableExecuteHandle tableExecuteHandle,
+            ConnectorTableHandle updatedSourceTableHandle)
+    {
+        IcebergTableExecuteHandle executeHandle = (IcebergTableExecuteHandle) tableExecuteHandle;
+        IcebergTableHandle table = (IcebergTableHandle) updatedSourceTableHandle;
+        switch (executeHandle.getProcedureId()) {
+            case OPTIMIZE:
+                return beginOptimize(session, executeHandle, table);
+        }
+        throw new IllegalArgumentException("Unknown procedure '" + executeHandle.getProcedureId() + "'");
+    }
+
+    private BeginTableExecuteResult<ConnectorTableExecuteHandle, ConnectorTableHandle> beginOptimize(
+            ConnectorSession session,
+            IcebergTableExecuteHandle executeHandle,
+            IcebergTableHandle table)
+    {
+        IcebergOptimizeHandle optimizeHandle = (IcebergOptimizeHandle) executeHandle.getProcedureHandle();
+        Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
+
+        verify(transaction == null, "transaction already set");
+        transaction = icebergTable.newTransaction();
+
+        return new BeginTableExecuteResult<>(
+                executeHandle.withProcedureHandle(optimizeHandle.withSnapshotId(Optional.ofNullable(icebergTable.currentSnapshot()).map(Snapshot::snapshotId))),
+                table.forOptimize(true, optimizeHandle.getMaxScannedFileSize()));
```

Comment:
I would skip passing `true` here. But I recall I had similar comment somewhere and you did not want any logic in Handle class, right?

Reply:
Right. I want IcebergTableHandle to be dumb state carrier, and logic kept here.